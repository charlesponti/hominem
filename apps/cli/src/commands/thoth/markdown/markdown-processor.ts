import { logger } from '@ponti/utils/logger'
import type { TextAnalysis } from '@ponti/utils/nlp'
import { getDatesFromText } from '@ponti/utils/time'
import * as cheerio from 'cheerio'
import * as fs from 'node:fs/promises'
import rehypeStringify from 'rehype-stringify'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import { unified } from 'unified'
import { detectTask, normalizeWhitespace, taskRegex } from './utils'

interface GetProcessedEntryParams {
  $: ReturnType<typeof cheerio.load>
  elem: ReturnType<ReturnType<typeof cheerio.load>>[0]
  entry: ProcessedMarkdownFileEntry
  section: string | null
  tag: string
}

export interface EntryContent {
  tag: string
  text: string
  section: string | null
  isTask?: boolean
  isComplete?: boolean
  textAnalysis?: TextAnalysis
  subentries?: EntryContent[]
}

export interface ProcessedMarkdownFileEntry {
  content: EntryContent[]
  date: string | undefined
  filename: string
  heading: string
  frontmatter?: Record<string, unknown>
}

export interface ProcessedMarkdownFile {
  entries: ProcessedMarkdownFileEntry[]
}

export class MarkdownProcessor {
  async processFileWithAst(filepath: string): Promise<ProcessedMarkdownFile> {
    // Check that the file exists
    if (!(await fs.stat(filepath)).isFile()) {
      throw new Error(`File not found: ${filepath}`)
    }

    const content = await fs.readFile(filepath, 'utf-8')
    const { result } = await this.convertMarkdownToJSON(content, filepath)
    return result
  }

  processFrontmatter(content: string): { content: string; frontmatter?: Record<string, unknown> } {
    // Find the frontmatter content at the start of the file
    const frontmatterMatch = content.match(/^---\n([\s\S]*?)\n---\n/)
    let frontmatter: Record<string, unknown> | undefined
    let processableContent = content

    if (frontmatterMatch) {
      try {
        // Parse the frontmatter content (assuming YAML format)
        const frontmatterContent = frontmatterMatch[1]
        // For now, we'll do a simple key-value extraction
        frontmatter = Object.fromEntries(
          frontmatterContent
            .split('\n')
            .filter((line) => line.includes(':'))
            .map((line) => {
              const [key, ...valueParts] = line.split(':')
              const value = valueParts.join(':').trim()
              // Remove quotes if present
              return [key.trim(), value.replace(/^['"]|['"]$/g, '')]
            })
        )
        // Remove the frontmatter from the content
        processableContent = content.slice(frontmatterMatch[0].length)
      } catch (error) {
        logger.warn('Failed to parse frontmatter:', error)
      }
    }

    return { content: processableContent, frontmatter }
  }

  async convertMarkdownToHTML(content: string) {
    const file = await unified()
      .use(remarkParse)
      .use(remarkRehype)
      .use(rehypeStringify)
      .process(content)

    return cheerio.load(String(file))
  }

  convertMarkdownToAST(content: string) {
    const processor = unified().use(remarkParse)
    return processor.parse(content)
  }

  // Helper method to create or retrieve an entry
  private ensureEntry(
    filename: string,
    heading: string | undefined,
    frontmatter?: Record<string, unknown>,
    entries: ProcessedMarkdownFileEntry[] = []
  ): ProcessedMarkdownFileEntry {
    // If no entry exists or we're at a new heading, create a new entry
    const entry: ProcessedMarkdownFileEntry = {
      date: undefined,
      filename,
      heading: heading || filename.split('/').pop() || '',
      content: [],
      frontmatter,
    }

    entries.push(entry)
    return entry
  }

  async convertMarkdownToJSON(
    markdownContent: string,
    filename: string
  ): Promise<{ result: ProcessedMarkdownFile; html: string }> {
    const result: ProcessedMarkdownFile = { entries: [] }

    // Extract frontmatter if present
    const { content: processableContent, frontmatter } = this.processFrontmatter(markdownContent)

    // Convert markdown to HTML for fancy querying
    const $ = await this.convertMarkdownToHTML(processableContent)

    let currentHeading: string | undefined
    let currentEntry: ProcessedMarkdownFileEntry | null = null

    const elements = $('body').children().toArray()
    for (const elem of elements) {
      const tag = elem.tagName.toLowerCase()
      const text = $(elem).text().trim()
      if (!text) continue

      const { taskText } = detectTask(text)

      const processedText =
        taskText ||
        text
          .replace(/^(-|\*)\s+/, '')
          .replace(taskRegex, '')
          .trim()

      if (/^h[1-6]$/.test(tag)) {
        // New entry on headings
        currentHeading = text
        currentEntry = this.ensureEntry(filename, currentHeading, frontmatter, result.entries)
        const { dates, fullDate } = getDatesFromText(text)
        currentEntry.date = dates.length > 0 ? dates[0].start.split('T')[0] : fullDate
      }

      if (tag === 'paragraph') {
        const previousContent = this.getPreviousEntry(currentEntry)

        if (previousContent && previousContent.tag === 'paragraph') {
          // Add to previous paragraph
          previousContent.text += `\n ${processedText}`
        }

        continue
      }

      if (!currentEntry) {
        currentEntry = this.ensureEntry(filename, currentHeading, frontmatter, result.entries)
      }

      if (tag === 'ul' || tag === 'ol') {
        const listItems = $(elem).find('> li').toArray()
        for (const li of listItems) {
          await this.processListItem({
            $,
            elem: li,
            entry: currentEntry,
            section: currentHeading ? currentHeading.toLowerCase() : null,
            tag,
          })
        }

        continue
      }

      const processedContent = await this.getProcessedEntry({
        $,
        elem,
        entry: currentEntry,
        section: currentHeading ? currentHeading.toLowerCase() : null,
        tag,
      })
      if (processedContent) {
        currentEntry.content.push(processedContent)
      }
    }

    if (result.entries.length === 0 && frontmatter) {
      this.ensureEntry(filename, undefined, frontmatter, result.entries)
    }

    return { result, html: $.html() }
  }

  async getProcessedEntry(params: GetProcessedEntryParams): Promise<EntryContent | undefined> {
    const { $, elem, section, tag } = params
    const text = $(elem).contents().first().text().trim()
    if (!text) return
    const normalizedText = normalizeWhitespace(text)
    const { isTask, isComplete, taskText } = detectTask(normalizedText)

    return {
      tag,
      text: taskText || normalizedText,
      section,
      isTask,
      isComplete,
      subentries: [],
    }
  }

  async processListItem(params: GetProcessedEntryParams) {
    const { $, elem, entry } = params
    const processedContent = await this.getProcessedEntry(params)
    if (!processedContent) return

    entry.content.push(processedContent)

    // Refactored nested lists recursion
    const subentries = await this.processNestedLists($, elem, entry, processedContent.tag)
    if (subentries.length > 0) {
      processedContent.subentries = subentries
    }
  }

  private async processNestedLists(
    $: ReturnType<typeof cheerio.load>,
    liElem: ReturnType<ReturnType<typeof cheerio.load>>[0],
    entry: ProcessedMarkdownFileEntry,
    tag: string
  ): Promise<EntryContent[]> {
    const subentries: EntryContent[] = []

    // Process immediate nested lists (ul or ol) inside the current li element
    const nestedLists = $(liElem).find('> ul, > ol').toArray()
    for (const nestedList of nestedLists) {
      // Use the text of the current li as section
      const parentText = $(liElem).contents().first().text().trim()
      const nestedListItems = $(nestedList).find('> li').toArray()
      for (const nestedListItem of nestedListItems) {
        const nestedContent = await this.getProcessedEntry({
          $,
          elem: nestedListItem,
          entry,
          section: parentText,
          tag,
        })
        if (nestedContent) {
          // Process deeper nested lists recursively
          nestedContent.subentries = await this.processNestedLists($, nestedListItem, entry, tag)
          subentries.push(nestedContent)
        }
      }
    }
    return subentries
  }

  getPreviousEntry(entry: ProcessedMarkdownFileEntry | null): EntryContent | undefined {
    if (!entry) return undefined
    return entry.content[entry.content.length - 1]
  }
}
