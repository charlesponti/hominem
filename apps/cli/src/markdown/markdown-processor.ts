import * as mdast from 'mdast-util-to-string'
import remarkParse from 'remark-parse'
import { unified } from 'unified'
import type { Node } from 'unist'
import type { Note, ProcessedContent } from './types'
import { getDateFromText, sanitizeText } from './utils'

interface MarkdownNode extends Node {
  type: string
  children?: MarkdownNode[]
  value?: string
  depth?: number
}

export class MarkdownProcessor {
  private currentHeading: string | undefined
  private content: ProcessedContent

  constructor() {
    this.content = {
      headings: [],
      paragraphs: [],
      bulletPoints: [],
      others: [],
    }
  }

  async processFile(filepath: string, content: string): Promise<ProcessedContent> {
    const tree = await unified().use(remarkParse).parse(content)

    this.traverseNodes(tree as MarkdownNode, filepath)
    return this.content
  }

  private traverseNodes(node: MarkdownNode, filename: string): void {
    if (node.type === 'heading') {
      const text = mdast.toString(node)
      this.currentHeading = sanitizeText(text)
      this.content.headings.push(this.currentHeading)
    }

    if (node.type === 'paragraph') {
      const text = mdast.toString(node)
      const { fullDate } = getDateFromText(text)

      this.content.paragraphs.push({
        file: filename,
        heading: this.currentHeading,
        text: sanitizeText(text),
        tag: 'paragraph',
        date: fullDate,
      })
    }

    if (node.type === 'list') {
      for (const item of node.children as MarkdownNode[]) {
        if (item.type === 'listItem') {
          const text = mdast.toString(item)
          const { fullDate } = getDateFromText(text)

          this.content.bulletPoints.push({
            file: filename,
            heading: this.currentHeading,
            text: sanitizeText(text),
            tag: 'bullet_point',
            date: fullDate,
          })
        }
      }
    }

    // Traverse children
    if (node.children) {
      for (const child of node.children) {
        this.traverseNodes(child, filename)
      }
    }
  }
}
