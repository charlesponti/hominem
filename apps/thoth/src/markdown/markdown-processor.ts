import * as cheerio from 'cheerio'
import rehypeStringify from 'rehype-stringify'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import { unified } from 'unified'
import { detectTask, normalizeWhitespace, taskRegex } from './utils'
import { getDatesFromText } from './time-utils'

interface GetProcessedEntryParams {
  $: ReturnType<typeof cheerio.load>
  elem: ReturnType<ReturnType<typeof cheerio.load>>[0]
  entry: ProcessedMarkdownFileEntry
  section: string | null
  tag: string
}

export interface TextAnalysis {
  entities?: Record<string, any>
  topics?: string[]
  sentiment?: {
    score: number
    label: string
  }
  summary?: string
  keywords?: string[]
}

export interface EntryContent {
  tag: string
  text: string
  section: string | null
  isTask?: boolean
  isComplete?: boolean
  textAnalysis?: TextAnalysis
  subentries?: EntryContent[]
}

export interface ProcessedMarkdownFileEntry {
  content: EntryContent[]
  date: string | undefined
  filename: string
  heading: string
  frontmatter?: Record<string, unknown>
}

export interface ProcessedMarkdownFile {
  entries: ProcessedMarkdownFileEntry[]
}

export class MarkdownProcessor {
  async getChunks(content: string, options?: { chunkSize?: number, chunkOverlap?: number }) {
    const separators = ['#', '##', '###', '####', '#####', '######']
    const chunkSize = options?.chunkSize || 2000
    const chunkOverlap = options?.chunkOverlap || 20
    
    // Split by the separators
    const chunks: string[] = []
    const lines = content.split('\n')
    let currentChunk = ''
    
    for (const line of lines) {
      const isHeading = separators.some(sep => line.trim().startsWith(sep + ' '))
      
      if (isHeading && currentChunk.length > 0) {
        chunks.push(currentChunk)
        currentChunk = line + '\n'
      } else {
        currentChunk += line + '\n'
        
        // Check if current chunk exceeds the chunk size
        if (currentChunk.length >= chunkSize) {
          chunks.push(currentChunk)
          // Keep some overlap for context
          const chunkLines = currentChunk.split('\n')
          const overlapLines = chunkLines.slice(Math.max(0, chunkLines.length - chunkOverlap))
          currentChunk = overlapLines.join('\n')
        }
      }
    }
    
    // Add the last chunk if not empty
    if (currentChunk.trim().length > 0) {
      chunks.push(currentChunk)
    }
    
    return chunks
  }

  processFrontmatter(content: string): { content: string; frontmatter?: Record<string, unknown> } {
    // Find the frontmatter content at the start of the file
    const frontmatterMatch = content.match(/^---\n([\s\S]*?)\n---\n/)
    let frontmatter: Record<string, unknown> | undefined
    let processableContent = content

    if (frontmatterMatch) {
      try {
        // Parse the frontmatter content (assuming YAML format)
        const frontmatterContent = frontmatterMatch[1]
        // For now, we'll do a simple key-value extraction
        frontmatter = Object.fromEntries(
          frontmatterContent
            .split('\n')
            .filter((line) => line.includes(':'))
            .map((line) => {
              const [key, ...valueParts] = line.split(':')
              const value = valueParts.join(':').trim()
              // Remove quotes if present
              return [key.trim(), value.replace(/^['"]|['"]$/g, '')]
            })
        )
        // Remove the frontmatter from the content
        processableContent = content.slice(frontmatterMatch[0].length)
      } catch (error) {
        console.warn('Failed to parse frontmatter:', error)
      }
    }

    return { content: processableContent, frontmatter }
  }

  async convertMarkdownToHTML(content: string) {
    const file = await unified()
      .use(remarkParse)
      .use(remarkRehype)
      .use(rehypeStringify)
      .process(content)

    return cheerio.load(String(file))
  }

  convertMarkdownToAST(content: string) {
    const processor = unified().use(remarkParse)
    return processor.parse(content)
  }

  // Helper method to create or retrieve an entry
  private ensureEntry(
    filename: string,
    heading: string | undefined,
    frontmatter?: Record<string, unknown>,
    entries: ProcessedMarkdownFileEntry[] = []
  ): ProcessedMarkdownFileEntry {
    // If no entry exists or we're at a new heading, create a new entry
    const entry: ProcessedMarkdownFileEntry = {
      date: undefined,
      filename,
      heading: heading || filename.split('/').pop() || '',
      content: [],
      frontmatter,
    }

    entries.push(entry)
    return entry
  }

  async convertMarkdownToJSON(
    markdownContent: string,
    filename: string
  ): Promise<{ result: ProcessedMarkdownFile; html: string }> {
    const result: ProcessedMarkdownFile = { entries: [] }

    // Extract frontmatter if present
    const { content: processableContent, frontmatter } = this.processFrontmatter(markdownContent)

    // Convert markdown to HTML for fancy querying
    const $ = await this.convertMarkdownToHTML(processableContent)

    let currentHeading: string | undefined
    let currentEntry: ProcessedMarkdownFileEntry | null = null

    const elements = $('body').children().toArray()
    for (const elem of elements) {
      const tag = elem.tagName.toLowerCase()
      const text = $(elem).text().trim()
      if (!text) continue

      const { taskText } = detectTask(text)

      const processedText =
        taskText ||
        text
          .replace(/^(-|\*)\s+/, '')
          .replace(taskRegex, '')
          .trim()

      if (/^h[1-6]$/.test(tag)) {
        // New entry on headings
        currentHeading = text
        currentEntry = this.ensureEntry(filename, currentHeading, frontmatter, result.entries)
        const { dates, fullDate } = getDatesFromText(text)
        currentEntry.date = dates.length > 0 ? dates[0].start.split('T')[0] : fullDate
      }

      if (tag === 'p') {
        const previousContent = this.getPreviousEntry(currentEntry)

        if (previousContent && previousContent.tag === 'p') {
          // Add to previous paragraph
          previousContent.text += `\n ${processedText}`
        }

        continue
      }

      if (!currentEntry) {
        currentEntry = this.ensureEntry(filename, currentHeading, frontmatter, result.entries)
      }

      if (tag === 'ul' || tag === 'ol') {
        const listItems = $(elem).find('> li').toArray()
        for (const li of listItems) {
          await this.processListItem({
            $,
            elem: li,
            entry: currentEntry,
            section: currentHeading ? currentHeading.toLowerCase() : null,
            tag,
          })
        }

        continue
      }

      const processedContent = await this.getProcessedEntry({
        $,
        elem,
        entry: currentEntry,
        section: currentHeading ? currentHeading.toLowerCase() : null,
        tag,
      })
      if (processedContent) {
        currentEntry.content.push(processedContent)
      }
    }

    if (result.entries.length === 0 && frontmatter) {
      this.ensureEntry(filename, undefined, frontmatter, result.entries)
    }

    return { result, html: $.html() }
  }

  async getProcessedEntry(params: GetProcessedEntryParams): Promise<EntryContent | undefined> {
    const { $, elem, section, tag } = params
    const text = $(elem).contents().first().text().trim()
    if (!text) return
    const normalizedText = normalizeWhitespace(text)
    const { isTask, isComplete, taskText } = detectTask(normalizedText)

    return {
      tag,
      text: taskText || normalizedText,
      section,
      isTask,
      isComplete,
      subentries: [],
    }
  }

  async processListItem(params: GetProcessedEntryParams) {
    const { $, elem, entry } = params
    const processedContent = await this.getProcessedEntry(params)
    if (!processedContent) return

    entry.content.push(processedContent)

    // Refactored nested lists recursion
    const subentries = await this.processNestedLists($, elem, entry, processedContent.tag)
    if (subentries.length > 0) {
      processedContent.subentries = subentries
    }
  }

  private async processNestedLists(
    $: ReturnType<typeof cheerio.load>,
    liElem: ReturnType<ReturnType<typeof cheerio.load>>[0],
    entry: ProcessedMarkdownFileEntry,
    tag: string
  ): Promise<EntryContent[]> {
    const subentries: EntryContent[] = []

    // Process immediate nested lists (ul or ol) inside the current li element
    const nestedLists = $(liElem).find('> ul, > ol').toArray()
    for (const nestedList of nestedLists) {
      // Use the text of the current li as section
      const parentText = $(liElem).contents().first().text().trim()
      const nestedListItems = $(nestedList).find('> li').toArray()
      for (const nestedListItem of nestedListItems) {
        const nestedContent = await this.getProcessedEntry({
          $,
          elem: nestedListItem,
          entry,
          section: parentText,
          tag,
        })
        if (nestedContent) {
          // Process deeper nested lists recursively
          nestedContent.subentries = await this.processNestedLists($, nestedListItem, entry, tag)
          subentries.push(nestedContent)
        }
      }
    }
    return subentries
  }

  getPreviousEntry(entry: ProcessedMarkdownFileEntry | null): EntryContent | undefined {
    if (!entry) return undefined
    return entry.content[entry.content.length - 1]
  }
}