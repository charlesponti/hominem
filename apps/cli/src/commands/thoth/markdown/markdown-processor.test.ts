import * as fs from 'node:fs/promises'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { MarkdownProcessor } from './markdown-processor'

vi.mock('node:fs/promises')
vi.mock('@ponti/utils/nlp')
// vi.mock('@ponti/utils/time', () => ({
//   getDatesFromText: vi.fn().mockReturnValue({ dates: [], fullDate: undefined }),
// }))

describe('MarkdownProcessor', () => {
  let processor: MarkdownProcessor

  beforeEach(() => {
    vi.resetAllMocks()
    processor = new MarkdownProcessor()

    // Mock fs.stat to always return { isFile: () => true } for file existence checks
    vi.mocked(fs.stat).mockResolvedValue({
      isFile: () => true,
      // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    } as any)
  })

  describe('processFrontmatter', () => {
    it('should extract frontmatter from markdown content', () => {
      const content = `---
title: Test Document
date: 2023-05-01
tags: test, markdown
---

# Heading 1

Some content here.`

      const result = processor.processFrontmatter(content)

      expect(result.frontmatter).toEqual({
        title: 'Test Document',
        date: '2023-05-01',
        tags: 'test, markdown',
      })

      expect(result.content).toContain('# Heading 1')
    })

    it('should handle markdown without frontmatter', () => {
      const content = '# Heading 1\n\nSome content here.'

      const result = processor.processFrontmatter(content)

      expect(result.frontmatter).toBeUndefined()
      expect(result.content).toBe(content)
    })
  })

  describe('convertMarkdownToJSON', () => {
    it('should process headings properly', async () => {
      const content =
        '# Heading 1\n\nParagraph under heading 1\n\n## Heading 2\n\nParagraph under heading 2'

      const result = await processor.convertMarkdownToJSON(content, 'test.md')

      expect(result.entries.length).toBe(2)
      expect(result.entries[0].heading).toBe('Heading 1')
      expect(result.entries[1].heading).toBe('Heading 2')
    })

    it('should process lists correctly', async () => {
      const content = `# Shopping List
      
- Apples
- Bananas
- Milk`

      const result = await processor.convertMarkdownToJSON(content, 'test.md')

      expect(result.entries.length).toBe(1)
      expect(result.entries[0].content.length).toBe(4)
      expect(result.entries[0].content[0].text).toBe('Shopping List')
      expect(result.entries[0].content[1].text).toBe('Apples')
      expect(result.entries[0].content[2].text).toBe('Bananas')
      expect(result.entries[0].content[3].text).toBe('Milk')
    })

    it('should detect tasks and their completion status', async () => {
      const content = `# Tasks
      
- [ ] Incomplete task
- [x] Complete task`

      const result = await processor.convertMarkdownToJSON(content, 'test.md')

      expect(result.entries[0].content[0].type).toBe('thought')
      expect(result.entries[0].content[0].text).toBe('Tasks')
      expect(result.entries[0].content[1].text).toBe('Incomplete task')
      expect(result.entries[0].content[1].type).toBe('task')
      expect(result.entries[0].content[1].isComplete).toBe(false)
      expect(result.entries[0].content[2].text).toBe('Complete task')
      expect(result.entries[0].content[2].type).toBe('task')
      expect(result.entries[0].content[2].isComplete).toBe(true)
    })

    it('should handle nested lists properly', async () => {
      const content = `# Nested List
      
- Fruits:
  - Apples
  - Bananas
- Vegetables:
  - Carrots
  - Broccoli`

      const result = await processor.convertMarkdownToJSON(content, 'test.md')

      // Check that we have exactly 2 top level items (Fruits and Vegetables)
      expect(result.entries[0].content.length).toBe(3)

      // Check first item (Fruits)
      expect(result.entries[0].content[1].text).toBe('Fruits:')
      expect(result.entries[0].content[1].subItems?.length).toBe(2)
      expect(result.entries[0].content[1].subItems?.[0].text).toBe('Apples')
      expect(result.entries[0].content[1].subItems?.[1].text).toBe('Bananas')

      // Check second item (Vegetables)
      expect(result.entries[0].content[2].text).toBe('Vegetables:')
      expect(result.entries[0].content[2].subItems?.length).toBe(2)
      expect(result.entries[0].content[2].subItems?.[0].text).toBe('Carrots')
      expect(result.entries[0].content[2].subItems?.[1].text).toBe('Broccoli')
    })

    it('should extract metadata from content', async () => {
      const content = `# Meeting Notes
      
Meeting with John Smith in New York about #project planning.`

      const result = await processor.convertMarkdownToJSON(content, 'test.md')

      expect(result.entries[0].content[0].text).toBe('Meeting Notes')
      expect(result.entries[0].content[1].metadata?.people).toContain('John Smith')
      expect(result.entries[0].content[1].section).toContain('meeting notes')
      expect(result.entries[0].content[1].metadata?.location).toBe('New York')
      expect(result.entries[0].content[1].metadata?.tags).toContain('project')
    })

    it('should correctly process hierarchical lists with personal reflections', async () => {
      const markdown = `## personal
- **Perfectionism**
- **Overthinking, analysis paralysis, fear of failure**
  - decrease my playfulness.
  - decreases poor decision-making
  - decreases action-taking
  - reinforces personal beliefs instead of increasing objective truth
- **Fear of judgement**
  - desire to be seen as correct instead of desire to be effective
  - many judgments should be ignored because they lack sincerity`

      const result = await processor.convertMarkdownToJSON(markdown, 'personal-reflections.md')

      // Test the basic structure
      expect(result.entries).toHaveLength(1)
      expect(result.entries[0].heading).toBe('personal')
      expect(result.entries[0].content).toHaveLength(3) // Three main bullet points

      // Test the main bullet points
      const [perfectionism, overthinking, fearOfJudgement] = result.entries[0].content

      // Test Perfectionism entry
      expect(perfectionism.text).toBe('Perfectionism')
      expect(perfectionism.type).toBe('thought')
      expect(perfectionism.tag).toBe('listItem')

      // Test Overthinking entry
      expect(overthinking.text).toBe('Overthinking, analysis paralysis, fear of failure')
      expect(overthinking.type).toBe('thought')
      expect(overthinking.subItems).toHaveLength(4)
      expect(overthinking.subItems?.map((item) => item.text)).toEqual([
        'decrease my playfulness.',
        'decreases poor decision-making',
        'decreases action-taking',
        'reinforces personal beliefs instead of increasing objective truth',
      ])

      // Test Fear of judgement entry
      expect(fearOfJudgement.text).toBe('Fear of judgement')
      expect(fearOfJudgement.type).toBe('thought')
      expect(fearOfJudgement.subItems).toHaveLength(2)
      expect(fearOfJudgement.subItems?.map((item) => item.text)).toEqual([
        'desire to be seen as correct instead of desire to be effective',
        'many judgments should be ignored because they lack sincerity',
      ])
    })
  })

  describe('processFileWithAst', () => {
    it('should read file and process its content', async () => {
      const fileContent = '# Test Heading\n\nTest content'
      vi.mocked(fs.readFile).mockResolvedValue(fileContent)

      const result = await processor.processFileWithAst('test.md')

      expect(fs.readFile).toHaveBeenCalledWith('test.md', 'utf-8')
      expect(result.entries.length).toBe(1)
      expect(result.entries[0].heading).toBe('Test Heading')
    })

    it('should throw an error for non-existent files', async () => {
      vi.mocked(fs.stat).mockResolvedValue({
        isFile: () => false,
        // biome-ignore lint/suspicious/noExplicitAny: <explanation>
      } as any)

      await expect(processor.processFileWithAst('nonexistent.md')).rejects.toThrow('File not found')
    })
  })
})
