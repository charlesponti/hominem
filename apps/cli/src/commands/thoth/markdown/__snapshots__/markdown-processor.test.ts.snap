// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MarkdownProcessor > convertMarkdownToJSON > should correctly process hierarchical lists with personal reflections 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "personal",
          "subentries": [],
          "tag": "h2",
          "text": "personal",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "personal",
          "subentries": [],
          "tag": "ul",
          "text": "Perfectionism",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "personal",
          "subentries": [
            {
              "isComplete": false,
              "isTask": false,
              "section": "Overthinking, analysis paralysis, fear of failure",
              "subentries": [],
              "tag": "ul",
              "text": "decrease my playfulness.",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Overthinking, analysis paralysis, fear of failure",
              "subentries": [],
              "tag": "ul",
              "text": "decreases poor decision-making",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Overthinking, analysis paralysis, fear of failure",
              "subentries": [],
              "tag": "ul",
              "text": "decreases action-taking",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Overthinking, analysis paralysis, fear of failure",
              "subentries": [],
              "tag": "ul",
              "text": "reinforces personal beliefs instead of increasing objective truth",
            },
          ],
          "tag": "ul",
          "text": "Overthinking, analysis paralysis, fear of failure",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "personal",
          "subentries": [
            {
              "isComplete": false,
              "isTask": false,
              "section": "Fear of judgement",
              "subentries": [],
              "tag": "ul",
              "text": "desire to be seen as correct instead of desire to be effective",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Fear of judgement",
              "subentries": [],
              "tag": "ul",
              "text": "many judgments should be ignored because they lack sincerity",
            },
          ],
          "tag": "ul",
          "text": "Fear of judgement",
        },
      ],
      "date": undefined,
      "filename": "personal-reflections.md",
      "frontmatter": undefined,
      "heading": "personal",
    },
  ],
}
`;

exports[`MarkdownProcessor > convertMarkdownToJSON > should detect tasks and their completion status 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "tasks",
          "subentries": [],
          "tag": "h1",
          "text": "Tasks",
        },
        {
          "isComplete": false,
          "isTask": true,
          "section": "tasks",
          "subentries": [],
          "tag": "ul",
          "text": "Incomplete task",
        },
        {
          "isComplete": true,
          "isTask": true,
          "section": "tasks",
          "subentries": [],
          "tag": "ul",
          "text": "Complete task",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Tasks",
    },
  ],
}
`;

exports[`MarkdownProcessor > convertMarkdownToJSON > should extract metadata from content 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "meeting notes",
          "subentries": [],
          "tag": "h1",
          "text": "Meeting Notes",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "meeting notes",
          "subentries": [],
          "tag": "p",
          "text": "Meeting with John Smith in New York about #project planning.",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Meeting Notes",
    },
  ],
}
`;

exports[`MarkdownProcessor > convertMarkdownToJSON > should handle nested lists properly 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "nested list",
          "subentries": [],
          "tag": "h1",
          "text": "Nested List",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "nested list",
          "subentries": [
            {
              "isComplete": false,
              "isTask": false,
              "section": "Fruits:",
              "subentries": [],
              "tag": "ul",
              "text": "Apples",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Fruits:",
              "subentries": [],
              "tag": "ul",
              "text": "Bananas",
            },
          ],
          "tag": "ul",
          "text": "Fruits:",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "nested list",
          "subentries": [
            {
              "isComplete": false,
              "isTask": false,
              "section": "Vegetables:",
              "subentries": [],
              "tag": "ul",
              "text": "Carrots",
            },
            {
              "isComplete": false,
              "isTask": false,
              "section": "Vegetables:",
              "subentries": [],
              "tag": "ul",
              "text": "Broccoli",
            },
          ],
          "tag": "ul",
          "text": "Vegetables:",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Nested List",
    },
  ],
}
`;

exports[`MarkdownProcessor > convertMarkdownToJSON > should process headings properly 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "heading 1",
          "subentries": [],
          "tag": "h1",
          "text": "Heading 1",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "heading 1",
          "subentries": [],
          "tag": "p",
          "text": "Paragraph under heading 1",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Heading 1",
    },
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "heading 2",
          "subentries": [],
          "tag": "h2",
          "text": "Heading 2",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "heading 2",
          "subentries": [],
          "tag": "p",
          "text": "Paragraph under heading 2",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Heading 2",
    },
  ],
}
`;

exports[`MarkdownProcessor > convertMarkdownToJSON > should process lists correctly 1`] = `
{
  "entries": [
    {
      "content": [
        {
          "isComplete": false,
          "isTask": false,
          "section": "shopping list",
          "subentries": [],
          "tag": "h1",
          "text": "Shopping List",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "shopping list",
          "subentries": [],
          "tag": "ul",
          "text": "Apples",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "shopping list",
          "subentries": [],
          "tag": "ul",
          "text": "Bananas",
        },
        {
          "isComplete": false,
          "isTask": false,
          "section": "shopping list",
          "subentries": [],
          "tag": "ul",
          "text": "Milk",
        },
      ],
      "date": undefined,
      "filename": "test.md",
      "frontmatter": undefined,
      "heading": "Shopping List",
    },
  ],
}
`;
